crates/symbolicator-service/src/caching/mod.rs

```diff
@@ -0,0 +1,233 @@
+//! This module includes all the code that deals with the different layers of caching, our central
+//! [`CacheError`] type, and contains an explanation of how all this works and why it exists.
+//!
+//! ## Cache Layers
+//!
+//! Symbolicator has a multi-layered caching architecture, consisting of the following layers:
+//!
+//! - An in-memory caching layer which is currently used for request coalescing
+//!   (deduplicating concurrent accesses).
+//! - A file-system layer that persists the results of downloads and computations to the file system,
+//!   and also persists errors happening during those.
+//! - A shared-cache layer which is being backed by a shared GCS bucket, to more evenly distribute
+//!   the load to multiple Symbolicator instances, and to help the fresh startup path without any
+//!   file-system caches available. The shared-cache layer does not persist errors, but only
+//!   "immutable" cache results.
+//!
+//! A cache request goes through the following steps:
+//! - First, it goes through the in-memory layer.
+//! - On miss, it will try to load the cache item from the file-system.
+//! - On miss, it will try the shared-cache next.
+//! - On miss, it will finally generate a fresh item, by downloading or doing a computation,
+//!   possibly requesting another cached item.
+//!
+//! ### Metrics
+//!
+//! TODO
+//!
+//! ### Configuration
+//!
+//! The request-coalescing part of the in-memory caching layer is always active. Additional
+//! configuration for the in-memory cache may be added in the future.
+//!
+//! The rest of the caching infrastructure is gated by the [`Config::cache_dir`] option. If no
+//! `cache_dir` is configured, a request is computed directly after the in-memory cache.
+//!
+//! [`Config::caches`] is responsible for the configuration of the file-system caches.
+//! It is divided into the categories "downloaded" and "derived". Both categories have settings
+//! related to cache expiration. Additionally, they allow configuring a limit on concurrent lazy
+//! re-downloads and re-computation. The limit applies to the whole category at once.
+//! See the section on [`CacheVersions`] for more details.
+//!
+//! The `retry_X_after` options specify a time-to-live after which the cache expires and will be
+//! re-computed. The `max_unused_for` option is rather a time-to-idle time, after which the item
+//! will be evicted. File-system `mtime` is used to check for these. Cache items that are in use
+//! will have their `mtime` updated once an hour to keep them from expiring.
+//!
+//! The "downloaded" category defaults to keeping entries alive for up to 24 hours, will retry
+//! "missing" items every hour, and "malformed" items every 24 hours.
+//! The "derived" category will keep entries alive for up to 7 days, and will also retry "missing"
+//! entries every hour, and "malformed" entries every 24 hours.
+//!
+//! A "successful" entry is considered immutable (with one exception) and it will be reused
```

Comment:
What is the exception?

Reply:
the current `should_load` code that throws away caches depending on symcache markers. That will go away once we properly switched over to fully unique / immutable keys in #983 . It might be good to call that out, also in the section about `CacheKey`.